## Preprocessing

In [1]:
# Importing our dependencies
from sklearn.model_selection import train_test_split
from sklearn.preprocessing import StandardScaler
import pandas as pd
import tensorflow as tf
import pandas as pd 

In [2]:
# Importing and reading the charity_data.csv.
application_df = pd.read_csv("https://static.bc-edx.com/data/dla-1-2/m21/lms/starter/charity_data.csv")
application_df_copy = application_df.copy()
application_df.head()

EIN                                      NAME APPLICATION_TYPE  \
0  10520599              BLUE KNIGHTS MOTORCYCLE CLUB              T10   
1  10531628    AMERICAN CHESAPEAKE CLUB CHARITABLE TR               T3   
2  10547893        ST CLOUD PROFESSIONAL FIREFIGHTERS               T5   
3  10553066            SOUTHSIDE ATHLETIC ASSOCIATION               T3   
4  10556103  GENETIC RESEARCH INSTITUTE OF THE DESERT               T3   

        AFFILIATION CLASSIFICATION      USE_CASE  ORGANIZATION  STATUS  \
0       Independent          C1000    ProductDev   Association       1   
1       Independent          C2000  Preservation  Co-operative       1   
2  CompanySponsored          C3000    ProductDev   Association       1   
3  CompanySponsored          C2000  Preservation         Trust       1   
4       Independent          C1000     Heathcare         Trust       1   

      INCOME_AMT SPECIAL_CONSIDERATIONS  ASK_AMT  IS_SUCCESSFUL  
0              0                      N     5000              1  
1         1-9999                      N   108590              1  
2              0                      N     5000              0  
3    10000-24999                      N     6692              1  
4  100000-499999                      N   142590              1

In [3]:
# Dropping the non-beneficial ID columns, 'EIN' and 'NAME'.
application_df = application_df.drop(columns=['EIN', 'NAME'], axis=1)

In [4]:
# Determining the number of unique values in each column.
application_df.nunique()

APPLICATION_TYPE            17
AFFILIATION                  6
CLASSIFICATION              71
USE_CASE                     5
ORGANIZATION                 4
STATUS                       2
INCOME_AMT                   9
SPECIAL_CONSIDERATIONS       2
ASK_AMT                   8747
IS_SUCCESSFUL                2
dtype: int64

In [5]:
# Looking at APPLICATION_TYPE value counts for binning
application_df['APPLICATION_TYPE'].value_counts()

T3     27037
T4      1542
T6      1216
T5      1173
T19     1065
T8       737
T7       725
T10      528
T9       156
T13       66
T12       27
T2        16
T25        3
T14        3
T29        2
T15        2
T17        1
Name: APPLICATION_TYPE, dtype: int64

In [6]:
# Unique APPLICATION_TYPE values to ease binning
application_df['APPLICATION_TYPE'].value_counts()[application_df['APPLICATION_TYPE'].value_counts() < 500]

T9     156
T13     66
T12     27
T2      16
T25      3
T14      3
T29      2
T15      2
T17      1
Name: APPLICATION_TYPE, dtype: int64

In [7]:
# Choosing a cutoff value and create a list of application types to be replaced
application_types_to_replace = ['T9', 'T13', 'T12', 'T2', 'T25', 'T14', 'T29', 'T15', 'T17']

# Replacing in dataframe
for app in application_types_to_replace:
    application_df['APPLICATION_TYPE'] = application_df['APPLICATION_TYPE'].replace(app,"Other")

# Checking to make sure binning was successful
application_df['APPLICATION_TYPE'].value_counts()

T3       27037
T4        1542
T6        1216
T5        1173
T19       1065
T8         737
T7         725
T10        528
Other      276
Name: APPLICATION_TYPE, dtype: int64

In [8]:
# Looking at CLASSIFICATION value counts for binning
application_df['CLASSIFICATION'].value_counts()

C1000    17326
C2000     6074
C1200     4837
C3000     1918
C2100     1883
         ...  
C4120        1
C8210        1
C2561        1
C4500        1
C2150        1
Name: CLASSIFICATION, Length: 71, dtype: int64

In [9]:
# looking at CLASSIFICATION value counts >1
application_df['CLASSIFICATION'].value_counts()[application_df['CLASSIFICATION'].value_counts() > 1]

C1000    17326
C2000     6074
C1200     4837
C3000     1918
C2100     1883
C7000      777
C1700      287
C4000      194
C5000      116
C1270      114
C2700      104
C2800       95
C7100       75
C1300       58
C1280       50
C1230       36
C1400       34
C7200       32
C2300       32
C1240       30
C8000       20
C7120       18
C1500       16
C1800       15
C6000       15
C1250       14
C8200       11
C1238       10
C1278       10
C1235        9
C1237        9
C7210        7
C2400        6
C1720        6
C4100        6
C1257        5
C1600        5
C1260        3
C2710        3
C0           3
C3200        2
C1234        2
C1246        2
C1267        2
C1256        2
Name: CLASSIFICATION, dtype: int64

In [10]:
# Displaying list of unique classification to ease creating list for binning
application_df['CLASSIFICATION'].unique()

array(['C1000', 'C2000', 'C3000', 'C1200', 'C2700', 'C7000', 'C7200',
       'C1700', 'C4000', 'C7100', 'C2800', 'C6000', 'C2100', 'C1238',
       'C5000', 'C7120', 'C1800', 'C4100', 'C1400', 'C1270', 'C2300',
       'C8200', 'C1500', 'C7210', 'C1300', 'C1230', 'C1280', 'C1240',
       'C2710', 'C2561', 'C1250', 'C8000', 'C1245', 'C1260', 'C1235',
       'C1720', 'C1257', 'C4500', 'C2400', 'C8210', 'C1600', 'C1278',
       'C1237', 'C4120', 'C2170', 'C1728', 'C1732', 'C2380', 'C1283',
       'C1570', 'C2500', 'C1267', 'C3700', 'C1580', 'C2570', 'C1256',
       'C1236', 'C1234', 'C1246', 'C2190', 'C4200', 'C0', 'C3200',
       'C5200', 'C1370', 'C2600', 'C1248', 'C6100', 'C1820', 'C1900',
       'C2150'], dtype=object)

In [11]:
# Choosing a cutoff value and creating a list of classifications to be replaced
# using the variable name `classifications_to_replace`
classifications_to_replace = ['C2700', 'C7000', 'C7200',
       'C1700', 'C4000', 'C7100', 'C2800', 'C6000', 'C1238',
       'C5000', 'C7120', 'C1800', 'C4100', 'C1400', 'C1270', 'C2300',
       'C8200', 'C1500', 'C7210', 'C1300', 'C1230', 'C1280', 'C1240',
       'C2710', 'C2561', 'C1250', 'C8000', 'C1245', 'C1260', 'C1235',
       'C1720', 'C1257', 'C4500', 'C2400', 'C8210', 'C1600', 'C1278',
       'C1237', 'C4120', 'C2170', 'C1728', 'C1732', 'C2380', 'C1283',
       'C1570', 'C2500', 'C1267', 'C3700', 'C1580', 'C2570', 'C1256',
       'C1236', 'C1234', 'C1246', 'C2190', 'C4200', 'C0', 'C3200',
       'C5200', 'C1370', 'C2600', 'C1248', 'C6100', 'C1820', 'C1900',
       'C2150']

# Replacing in dataframe
for cls in classifications_to_replace:
    application_df['CLASSIFICATION'] = application_df['CLASSIFICATION'].replace(cls,"Other")
    
# Checking to make sure binning was successful
application_df['CLASSIFICATION'].value_counts()

C1000    17326
C2000     6074
C1200     4837
Other     2261
C3000     1918
C2100     1883
Name: CLASSIFICATION, dtype: int64

In [12]:
# Getting data types of all columns
application_df.dtypes

APPLICATION_TYPE          object
AFFILIATION               object
CLASSIFICATION            object
USE_CASE                  object
ORGANIZATION              object
STATUS                     int64
INCOME_AMT                object
SPECIAL_CONSIDERATIONS    object
ASK_AMT                    int64
IS_SUCCESSFUL              int64
dtype: object

In [13]:
# Getting unique values and counts of all columns
application_df['AFFILIATION'].value_counts()

Independent         18480
CompanySponsored    15705
Family/Parent          64
National               33
Regional               13
Other                   4
Name: AFFILIATION, dtype: int64

In [14]:
application_df['USE_CASE'].value_counts()

Preservation     28095
ProductDev        5671
CommunityServ      384
Heathcare          146
Other                3
Name: USE_CASE, dtype: int64

In [15]:
application_df['ORGANIZATION'].value_counts()

Trust           23515
Association     10255
Co-operative      486
Corporation        43
Name: ORGANIZATION, dtype: int64

In [16]:
application_df['INCOME_AMT'].value_counts()

0                24388
25000-99999       3747
100000-499999     3374
1M-5M              955
1-9999             728
10000-24999        543
10M-50M            240
5M-10M             185
50M+               139
Name: INCOME_AMT, dtype: int64

In [17]:
application_df['SPECIAL_CONSIDERATIONS'].value_counts()

N    34272
Y       27
Name: SPECIAL_CONSIDERATIONS, dtype: int64

In [18]:
application_df['STATUS'].value_counts()

1    34294
0        5
Name: STATUS, dtype: int64

In [19]:
# Converting categorical data to numeric with `pd.get_dummies`
application_df = pd.get_dummies(application_df,dtype=float)
application_df.head()

STATUS  ASK_AMT  IS_SUCCESSFUL  APPLICATION_TYPE_Other  \
0       1     5000              1                     0.0   
1       1   108590              1                     0.0   
2       1     5000              0                     0.0   
3       1     6692              1                     0.0   
4       1   142590              1                     0.0   

   APPLICATION_TYPE_T10  APPLICATION_TYPE_T19  APPLICATION_TYPE_T3  \
0                   1.0                   0.0                  0.0   
1                   0.0                   0.0                  1.0   
2                   0.0                   0.0                  0.0   
3                   0.0                   0.0                  1.0   
4                   0.0                   0.0                  1.0   

   APPLICATION_TYPE_T4  APPLICATION_TYPE_T5  APPLICATION_TYPE_T6  ...  \
0                  0.0                  0.0                  0.0  ...   
1                  0.0                  0.0                  0.0  ...   
2                  0.0                  1.0                  0.0  ...   
3                  0.0                  0.0                  0.0  ...   
4                  0.0                  0.0                  0.0  ...   

   INCOME_AMT_1-9999  INCOME_AMT_10000-24999  INCOME_AMT_100000-499999  \
0                0.0                     0.0                       0.0   
1                1.0                     0.0                       0.0   
2                0.0                     0.0                       0.0   
3                0.0                     1.0                       0.0   
4                0.0                     0.0                       1.0   

   INCOME_AMT_10M-50M  INCOME_AMT_1M-5M  INCOME_AMT_25000-99999  \
0                 0.0               0.0                     0.0   
1                 0.0               0.0                     0.0   
2                 0.0               0.0                     0.0   
3                 0.0               0.0                     0.0   
4                 0.0               0.0                     0.0   

   INCOME_AMT_50M+  INCOME_AMT_5M-10M  SPECIAL_CONSIDERATIONS_N  \
0              0.0                0.0                       1.0   
1              0.0                0.0                       1.0   
2              0.0                0.0                       1.0   
3              0.0                0.0                       1.0   
4              0.0                0.0                       1.0   

   SPECIAL_CONSIDERATIONS_Y  
0                       0.0  
1                       0.0  
2                       0.0  
3                       0.0  
4                       0.0  

[5 rows x 44 columns]

In [20]:
# Splitting preprocessed data into features and target arrays
y = application_df['IS_SUCCESSFUL'].values
X = application_df.drop(columns= 'IS_SUCCESSFUL').values

# Splitting the preprocessed data into a training and testing dataset
X_train, X_test, y_train, y_test = train_test_split(X, y, random_state=42, stratify=y)

In [21]:
# Creating a StandardScaler instance
scaler = StandardScaler()

# Fitting the StandardScaler
X_scaler = scaler.fit(X_train)

# Scaling the data
X_train_scaled = X_scaler.transform(X_train)
X_test_scaled = X_scaler.transform(X_test)

## Compile, Train and Evaluate the Model

## Model optimisation method 1
### Increasing number of layers and neurons

In [22]:
# Defining the model - deep neural net, i.e., the number of input features and hidden nodes for each layer.
number_input_features = len(X_train[0])

hidden_nodes_layer1 =  80
hidden_nodes_layer2 = 30
hidden_nodes_layer3 = 10

nn = tf.keras.models.Sequential()

# First hidden layer
nn.add(
    tf.keras.layers.Dense(units=hidden_nodes_layer1, input_dim=number_input_features, activation="relu")
)

# Second hidden layer
nn.add(tf.keras.layers.Dense(units=hidden_nodes_layer2, activation="relu"))

# Third hidden layer
nn.add(tf.keras.layers.Dense(units=hidden_nodes_layer3, activation="relu"))

# Output layer
nn.add(tf.keras.layers.Dense(units=1, activation="sigmoid"))

# Checking the structure of the model
nn.summary()

Model: "sequential"
_________________________________________________________________
 Layer (type)                Output Shape              Param #   
 dense (Dense)               (None, 80)                3520      
                                                                 
 dense_1 (Dense)             (None, 30)                2430      
                                                                 
 dense_2 (Dense)             (None, 10)                310       
                                                                 
 dense_3 (Dense)             (None, 1)                 11        
                                                                 
Total params: 6271 (24.50 KB)
Trainable params: 6271 (24.50 KB)
Non-trainable params: 0 (0.00 Byte)
_________________________________________________________________


In [23]:
# Compiling the model
nn.compile(loss="binary_crossentropy", optimizer="adam", metrics=["accuracy"])

In [24]:
# Train the model
fit_model = nn.fit(X_train_scaled,y_train,epochs=100)

Epoch 1/100
804/804 [==============================] - 4s 3ms/step - loss: 0.5683 - accuracy: 0.7218
Epoch 2/100
804/804 [==============================] - 2s 3ms/step - loss: 0.5545 - accuracy: 0.7294
Epoch 3/100
804/804 [==============================] - 3s 3ms/step - loss: 0.5513 - accuracy: 0.7310
Epoch 4/100
804/804 [==============================] - 3s 3ms/step - loss: 0.5494 - accuracy: 0.7339
Epoch 5/100
804/804 [==============================] - 2s 3ms/step - loss: 0.5495 - accuracy: 0.7340
Epoch 6/100
804/804 [==============================] - 3s 4ms/step - loss: 0.5478 - accuracy: 0.7352
Epoch 7/100
804/804 [==============================] - 3s 4ms/step - loss: 0.5477 - accuracy: 0.7348
Epoch 8/100
804/804 [==============================] - 2s 3ms/step - loss: 0.5471 - accuracy: 0.7353
Epoch 9/100
804/804 [==============================] - 2s 3ms/step - loss: 0.5459 - accuracy: 0.7357
Epoch 10/100
804/804 [==============================] - 2s 3ms/step - loss: 0.5459 - accura

804/804 [==============================] - 2s 3ms/step - loss: 0.5353 - accuracy: 0.7403
Epoch 82/100
804/804 [==============================] - 2s 3ms/step - loss: 0.5352 - accuracy: 0.7408
Epoch 83/100
804/804 [==============================] - 3s 3ms/step - loss: 0.5351 - accuracy: 0.7404
Epoch 84/100
804/804 [==============================] - 3s 3ms/step - loss: 0.5348 - accuracy: 0.7413
Epoch 85/100
804/804 [==============================] - 2s 3ms/step - loss: 0.5353 - accuracy: 0.7411
Epoch 86/100
804/804 [==============================] - 2s 3ms/step - loss: 0.5346 - accuracy: 0.7414
Epoch 87/100
804/804 [==============================] - 2s 3ms/step - loss: 0.5350 - accuracy: 0.7415
Epoch 88/100
804/804 [==============================] - 2s 3ms/step - loss: 0.5344 - accuracy: 0.7413
Epoch 89/100
804/804 [==============================] - 2s 3ms/step - loss: 0.5345 - accuracy: 0.7406
Epoch 90/100
804/804 [==============================] - 2s 3ms/step - loss: 0.5344 - accuracy: 

In [30]:
# Evaluating the model using the test data
model_loss, model_accuracy = nn.evaluate(X_test_scaled,y_test,verbose=2)
print(f"Loss: {model_loss}, Accuracy: {model_accuracy}")

268/268 - 1s - loss: 0.5569 - accuracy: 0.7234 - 814ms/epoch - 3ms/step
Loss: 0.556873083114624, Accuracy: 0.7233819365501404


### No change from initial model is observed. 
### Note: Tried many permutaitons and combinations by increasing neurons, increasing number of layers, increasing number of layers and neurons and kept one of the iterations here in the notebook. No observation had better accuracy.

## Model optimisation method 2
### Changing activation method and neurons

In [27]:
# Defining the model - deep neural net, i.e., the number of input features and hidden nodes for each layer.
number_input_features = len(X_train[0])

hidden_nodes_layer1 =  80
hidden_nodes_layer2 = 30
hidden_nodes_layer3 = 15

nn = tf.keras.models.Sequential()

# First hidden layer
nn.add(
    tf.keras.layers.Dense(units=hidden_nodes_layer1, input_dim=number_input_features, activation="relu")
)

# Second hidden layer
nn.add(tf.keras.layers.Dense(units=hidden_nodes_layer2, activation="relu"))

# Third hidden layer
nn.add(tf.keras.layers.Dense(units=hidden_nodes_layer3, activation="tanh"))

# Output layer
nn.add(tf.keras.layers.Dense(units=1, activation="sigmoid"))

# Checking the structure of the model
nn.summary()

Model: "sequential_1"
_________________________________________________________________
 Layer (type)                Output Shape              Param #   
 dense_4 (Dense)             (None, 80)                3520      
                                                                 
 dense_5 (Dense)             (None, 30)                2430      
                                                                 
 dense_6 (Dense)             (None, 15)                465       
                                                                 
 dense_7 (Dense)             (None, 1)                 16        
                                                                 
Total params: 6431 (25.12 KB)
Trainable params: 6431 (25.12 KB)
Non-trainable params: 0 (0.00 Byte)
_________________________________________________________________


In [28]:
# Compiling the model
nn.compile(loss="binary_crossentropy", optimizer="adam", metrics=["accuracy"])

In [29]:
# Train the model
fit_model = nn.fit(X_train_scaled,y_train,epochs=100)

Epoch 1/100
804/804 [==============================] - 4s 3ms/step - loss: 0.5683 - accuracy: 0.7203
Epoch 2/100
804/804 [==============================] - 2s 3ms/step - loss: 0.5539 - accuracy: 0.7303
Epoch 3/100
804/804 [==============================] - 2s 3ms/step - loss: 0.5512 - accuracy: 0.7316
Epoch 4/100
804/804 [==============================] - 2s 3ms/step - loss: 0.5497 - accuracy: 0.7317
Epoch 5/100
804/804 [==============================] - 2s 3ms/step - loss: 0.5486 - accuracy: 0.7339
Epoch 6/100
804/804 [==============================] - 2s 3ms/step - loss: 0.5479 - accuracy: 0.7322
Epoch 7/100
804/804 [==============================] - 2s 3ms/step - loss: 0.5466 - accuracy: 0.7319
Epoch 8/100
804/804 [==============================] - 2s 3ms/step - loss: 0.5466 - accuracy: 0.7329
Epoch 9/100
804/804 [==============================] - 2s 3ms/step - loss: 0.5454 - accuracy: 0.7356
Epoch 10/100
804/804 [==============================] - 2s 3ms/step - loss: 0.5454 - accura

804/804 [==============================] - 2s 3ms/step - loss: 0.5348 - accuracy: 0.7403
Epoch 82/100
804/804 [==============================] - 2s 3ms/step - loss: 0.5342 - accuracy: 0.7406
Epoch 83/100
804/804 [==============================] - 2s 3ms/step - loss: 0.5347 - accuracy: 0.7402
Epoch 84/100
804/804 [==============================] - 2s 3ms/step - loss: 0.5343 - accuracy: 0.7405
Epoch 85/100
804/804 [==============================] - 2s 3ms/step - loss: 0.5344 - accuracy: 0.7413
Epoch 86/100
804/804 [==============================] - 3s 3ms/step - loss: 0.5342 - accuracy: 0.7404
Epoch 87/100
804/804 [==============================] - 3s 4ms/step - loss: 0.5343 - accuracy: 0.7402
Epoch 88/100
804/804 [==============================] - 3s 4ms/step - loss: 0.5343 - accuracy: 0.7404
Epoch 89/100
804/804 [==============================] - 3s 4ms/step - loss: 0.5345 - accuracy: 0.7412
Epoch 90/100
804/804 [==============================] - 3s 3ms/step - loss: 0.5345 - accuracy: 

In [31]:
# Evaluating the model using the test data
model_loss, model_accuracy = nn.evaluate(X_test_scaled,y_test,verbose=2)
print(f"Loss: {model_loss}, Accuracy: {model_accuracy}")

268/268 - 1s - loss: 0.5569 - accuracy: 0.7234 - 562ms/epoch - 2ms/step
Loss: 0.556873083114624, Accuracy: 0.7233819365501404


### No change from previous model is observed at 72.3% accuracy

## Model optimisation method 3
### Dropping status and special considerations related columns

In [32]:
application_df = application_df.drop(columns=['STATUS','SPECIAL_CONSIDERATIONS_N','SPECIAL_CONSIDERATIONS_Y'], axis=1)

In [33]:
# Splitting preprocessed data into features and target arrays
y = application_df['IS_SUCCESSFUL'].values
X = application_df.drop(columns= 'IS_SUCCESSFUL').values

# Splitting the preprocessed data into a training and testing dataset
X_train, X_test, y_train, y_test = train_test_split(X, y, random_state=42, stratify=y)

# Creating a StandardScaler instance
scaler = StandardScaler()

# Fitting the StandardScaler
X_scaler = scaler.fit(X_train)

# Scaling the data
X_train_scaled = X_scaler.transform(X_train)
X_test_scaled = X_scaler.transform(X_test)

In [34]:
# Defining the model - deep neural net, i.e., the number of input features and hidden nodes for each layer.
number_input_features = len(X_train[0])

hidden_nodes_layer1 =  80
hidden_nodes_layer2 = 30

nn = tf.keras.models.Sequential()

# First hidden layer
nn.add(
    tf.keras.layers.Dense(units=hidden_nodes_layer1, input_dim=number_input_features, activation="relu")
)

# Second hidden layer
nn.add(tf.keras.layers.Dense(units=hidden_nodes_layer2, activation="relu"))

# Output layer
nn.add(tf.keras.layers.Dense(units=1, activation="sigmoid"))

# Checking the structure of the model
nn.summary()

Model: "sequential_2"
_________________________________________________________________
 Layer (type)                Output Shape              Param #   
 dense_8 (Dense)             (None, 80)                3280      
                                                                 
 dense_9 (Dense)             (None, 30)                2430      
                                                                 
 dense_10 (Dense)            (None, 1)                 31        
                                                                 
Total params: 5741 (22.43 KB)
Trainable params: 5741 (22.43 KB)
Non-trainable params: 0 (0.00 Byte)
_________________________________________________________________


In [35]:
# Compiling the model
nn.compile(loss="binary_crossentropy", optimizer="adam", metrics=["accuracy"])

In [36]:
# Train the model
fit_model = nn.fit(X_train_scaled,y_train,epochs=100)

Epoch 1/100
804/804 [==============================] - 3s 3ms/step - loss: 0.5693 - accuracy: 0.7228
Epoch 2/100
804/804 [==============================] - 2s 3ms/step - loss: 0.5550 - accuracy: 0.7299
Epoch 3/100
804/804 [==============================] - 2s 3ms/step - loss: 0.5515 - accuracy: 0.7313
Epoch 4/100
804/804 [==============================] - 2s 3ms/step - loss: 0.5512 - accuracy: 0.7333
Epoch 5/100
804/804 [==============================] - 2s 3ms/step - loss: 0.5492 - accuracy: 0.7324
Epoch 6/100
804/804 [==============================] - 2s 3ms/step - loss: 0.5483 - accuracy: 0.7332
Epoch 7/100
804/804 [==============================] - 2s 3ms/step - loss: 0.5469 - accuracy: 0.7344
Epoch 8/100
804/804 [==============================] - 2s 3ms/step - loss: 0.5471 - accuracy: 0.7355
Epoch 9/100
804/804 [==============================] - 2s 3ms/step - loss: 0.5461 - accuracy: 0.7347
Epoch 10/100
804/804 [==============================] - 3s 3ms/step - loss: 0.5455 - accura

804/804 [==============================] - 2s 3ms/step - loss: 0.5348 - accuracy: 0.7404
Epoch 82/100
804/804 [==============================] - 2s 3ms/step - loss: 0.5351 - accuracy: 0.7402
Epoch 83/100
804/804 [==============================] - 2s 3ms/step - loss: 0.5352 - accuracy: 0.7404
Epoch 84/100
804/804 [==============================] - 2s 3ms/step - loss: 0.5350 - accuracy: 0.7421
Epoch 85/100
804/804 [==============================] - 2s 3ms/step - loss: 0.5351 - accuracy: 0.7413
Epoch 86/100
804/804 [==============================] - 2s 3ms/step - loss: 0.5350 - accuracy: 0.7410
Epoch 87/100
804/804 [==============================] - 2s 3ms/step - loss: 0.5348 - accuracy: 0.7414
Epoch 88/100
804/804 [==============================] - 2s 3ms/step - loss: 0.5349 - accuracy: 0.7410
Epoch 89/100
804/804 [==============================] - 2s 3ms/step - loss: 0.5349 - accuracy: 0.7411
Epoch 90/100
804/804 [==============================] - 2s 3ms/step - loss: 0.5347 - accuracy: 

In [37]:
# Evaluating the model using the test data
model_loss, model_accuracy = nn.evaluate(X_test_scaled,y_test,verbose=2)
print(f"Loss: {model_loss}, Accuracy: {model_accuracy}")

268/268 - 1s - loss: 0.5661 - accuracy: 0.7245 - 857ms/epoch - 3ms/step
Loss: 0.566061794757843, Accuracy: 0.7245481014251709


### Again no change from previous model is observed at 72.4% accuracy

In [38]:
# Saving weights of first layer of this model just to check how weights are distributed
weights, biases = nn.layers[0].get_weights()
weights2 =pd.DataFrame(weights)
weights2.to_csv("Weights/checkweightmethod2.csv")

In [40]:
# Increasing number of epochs
fit_model = nn.fit(X_train_scaled,y_train,epochs=200)

Epoch 1/200
804/804 [==============================] - 2s 3ms/step - loss: 0.5327 - accuracy: 0.7423
Epoch 2/200
804/804 [==============================] - 2s 3ms/step - loss: 0.5329 - accuracy: 0.7420
Epoch 3/200
804/804 [==============================] - 2s 3ms/step - loss: 0.5321 - accuracy: 0.7425
Epoch 4/200
804/804 [==============================] - 2s 3ms/step - loss: 0.5323 - accuracy: 0.7420
Epoch 5/200
804/804 [==============================] - 2s 3ms/step - loss: 0.5321 - accuracy: 0.7424
Epoch 6/200
804/804 [==============================] - 2s 3ms/step - loss: 0.5323 - accuracy: 0.7430
Epoch 7/200
804/804 [==============================] - 2s 3ms/step - loss: 0.5325 - accuracy: 0.7421
Epoch 8/200
804/804 [==============================] - 2s 3ms/step - loss: 0.5324 - accuracy: 0.7427
Epoch 9/200
804/804 [==============================] - 2s 3ms/step - loss: 0.5319 - accuracy: 0.7420
Epoch 10/200
804/804 [==============================] - 2s 3ms/step - loss: 0.5320 - accura

804/804 [==============================] - 2s 3ms/step - loss: 0.5309 - accuracy: 0.7427
Epoch 82/200
804/804 [==============================] - 2s 3ms/step - loss: 0.5311 - accuracy: 0.7427
Epoch 83/200
804/804 [==============================] - 2s 3ms/step - loss: 0.5309 - accuracy: 0.7426
Epoch 84/200
804/804 [==============================] - 2s 3ms/step - loss: 0.5308 - accuracy: 0.7432
Epoch 85/200
804/804 [==============================] - 2s 3ms/step - loss: 0.5315 - accuracy: 0.7427
Epoch 86/200
804/804 [==============================] - 2s 3ms/step - loss: 0.5314 - accuracy: 0.7430
Epoch 87/200
804/804 [==============================] - 2s 3ms/step - loss: 0.5308 - accuracy: 0.7428
Epoch 88/200
804/804 [==============================] - 3s 3ms/step - loss: 0.5318 - accuracy: 0.7431
Epoch 89/200
804/804 [==============================] - 2s 3ms/step - loss: 0.5322 - accuracy: 0.7422
Epoch 90/200
804/804 [==============================] - 2s 3ms/step - loss: 0.5306 - accuracy: 

804/804 [==============================] - 2s 3ms/step - loss: 0.5299 - accuracy: 0.7433
Epoch 161/200
804/804 [==============================] - 2s 3ms/step - loss: 0.5309 - accuracy: 0.7427
Epoch 162/200
804/804 [==============================] - 2s 3ms/step - loss: 0.5300 - accuracy: 0.7430
Epoch 163/200
804/804 [==============================] - 2s 3ms/step - loss: 0.5301 - accuracy: 0.7433
Epoch 164/200
804/804 [==============================] - 3s 4ms/step - loss: 0.5298 - accuracy: 0.7431
Epoch 165/200
804/804 [==============================] - 3s 4ms/step - loss: 0.5304 - accuracy: 0.7431
Epoch 166/200
804/804 [==============================] - 3s 3ms/step - loss: 0.5305 - accuracy: 0.7425
Epoch 167/200
804/804 [==============================] - 3s 3ms/step - loss: 0.5304 - accuracy: 0.7430
Epoch 168/200
804/804 [==============================] - 3s 4ms/step - loss: 0.5302 - accuracy: 0.7428
Epoch 169/200
804/804 [==============================] - 2s 3ms/step - loss: 0.5305 - a

In [41]:
# Evaluating the model using the test data
model_loss, model_accuracy = nn.evaluate(X_test_scaled,y_test,verbose=2)
print(f"Loss: {model_loss}, Accuracy: {model_accuracy}")

268/268 - 1s - loss: 0.5981 - accuracy: 0.7234 - 556ms/epoch - 2ms/step
Loss: 0.598114550113678, Accuracy: 0.7233819365501404


### By increasing number of epochs to 200 we achieve 72.3% accuracy which is again the similar to the last models. Also to be noted that loss is around 53% in all above trials

## Model optimisation method 4

### Checking Name/EID column values for duplicates (reappearance over time) and if one of them needs to be included for creating our model

In [42]:
application_df_copy['EIN'].value_counts()[application_df_copy['EIN'].value_counts() >1]

Series([], Name: EIN, dtype: int64)

In [43]:
application_df_copy['NAME'].value_counts()[application_df_copy['NAME'].value_counts() >1]

PARENT BOOSTER USA INC                                            1260
TOPS CLUB INC                                                      765
UNITED STATES BOWLING CONGRESS INC                                 700
WASHINGTON STATE UNIVERSITY                                        492
AMATEUR ATHLETIC UNION OF THE UNITED STATES INC                    408
                                                                  ... 
VETERANS OF FOREIGH WARS OF THE UNITED STATES DEPT OF COLORADO       2
ETA PHI BETA SORORITY INC                                            2
POINT MAN INTERNATIONAL MINISTRIES                                   2
MULTI COMMUNITY DIVERSIFIED SERVICES INC                             2
VETERANS OF FOREIGN WARS AUXILIARY                                   2
Name: NAME, Length: 792, dtype: int64

### We can see that names of organisation are not unique like EIN which is the sole unique identifier for this data. We will drop only EIN column and try generating a model to see if it has any impact on accuracy of model when tested

In [44]:
# Dropping EIN column
application_df_copy = application_df_copy.drop(columns=['EIN'], axis=1)


In [46]:
# Binning names
# Choosing a cutoff value and creating a list of names to be replaced and replacing in dataframe
counts = application_df_copy['NAME'].value_counts()
names_to_replace = list(counts[counts <5].index)

In [47]:
for name in names_to_replace:
    application_df_copy['NAME'] = application_df_copy['NAME'].replace(name,"Other")
    
application_df_copy['NAME'].value_counts()

Other                                                                19803
PARENT BOOSTER USA INC                                                1260
TOPS CLUB INC                                                          765
UNITED STATES BOWLING CONGRESS INC                                     700
WASHINGTON STATE UNIVERSITY                                            492
                                                                     ...  
GUILDS OF THE SANTA FE OPERA INC                                         5
FAMILY MOTOR COACH ASSOCIATION                                           5
JAMESTOWN SOCIETY                                                        5
MOST WORSHIPFUL PRINCE HALL GRAND LODGE OF FREE & ACCEPTED MASONS        5
PTA HAWAII CONGRESS                                                      5
Name: NAME, Length: 403, dtype: int64

In [48]:
# Binning Application Types and classifications
# Choosing a cutoff value and creating a list of application types to be replaced and replacing in dataframe
application_types_to_replace = ['T9', 'T13', 'T12', 'T2', 'T25', 'T14', 'T29', 'T15', 'T17']

for app in application_types_to_replace:
    application_df_copy['APPLICATION_TYPE'] = application_df_copy['APPLICATION_TYPE'].replace(app,"Other")

# Choosing a cutoff value and creating a list of classifications to be replaced and replacing in dataframe
classifications_to_replace = ['C2700', 'C7000', 'C7200',
       'C1700', 'C4000', 'C7100', 'C2800', 'C6000', 'C1238',
       'C5000', 'C7120', 'C1800', 'C4100', 'C1400', 'C1270', 'C2300',
       'C8200', 'C1500', 'C7210', 'C1300', 'C1230', 'C1280', 'C1240',
       'C2710', 'C2561', 'C1250', 'C8000', 'C1245', 'C1260', 'C1235',
       'C1720', 'C1257', 'C4500', 'C2400', 'C8210', 'C1600', 'C1278',
       'C1237', 'C4120', 'C2170', 'C1728', 'C1732', 'C2380', 'C1283',
       'C1570', 'C2500', 'C1267', 'C3700', 'C1580', 'C2570', 'C1256',
       'C1236', 'C1234', 'C1246', 'C2190', 'C4200', 'C0', 'C3200',
       'C5200', 'C1370', 'C2600', 'C1248', 'C6100', 'C1820', 'C1900',
       'C2150']

for cls in classifications_to_replace:
    application_df_copy['CLASSIFICATION'] = application_df_copy['CLASSIFICATION'].replace(cls,"Other")

In [49]:
# Converting categorical data to numeric with `pd.get_dummies`
application_df_copy = pd.get_dummies(application_df_copy,dtype=float)
application_df_copy.head()

STATUS  ASK_AMT  IS_SUCCESSFUL  NAME_AACE INTERNATIONAL  \
0       1     5000              1                      0.0   
1       1   108590              1                      0.0   
2       1     5000              0                      0.0   
3       1     6692              1                      0.0   
4       1   142590              1                      0.0   

   NAME_ACE MENTOR PROGRAM OF AMERICA INC  NAME_ACTS MINISTRY  \
0                                     0.0                 0.0   
1                                     0.0                 0.0   
2                                     0.0                 0.0   
3                                     0.0                 0.0   
4                                     0.0                 0.0   

   NAME_ACTS MISSIONS  \
0                 0.0   
1                 0.0   
2                 0.0   
3                 0.0   
4                 0.0   

   NAME_AFRICAN-AMERICAN POSTAL LEAGUE UNITED FOR SUCCESS A-PLUS  \
0                                                0.0               
1                                                0.0               
2                                                0.0               
3                                                0.0               
4                                                0.0               

   NAME_AIR FORCE ASSOCIATION  NAME_ALABAMA FEDERATION OF WOMENS CLUBS  ...  \
0                         0.0                                      0.0  ...   
1                         0.0                                      0.0  ...   
2                         0.0                                      0.0  ...   
3                         0.0                                      0.0  ...   
4                         0.0                                      0.0  ...   

   INCOME_AMT_1-9999  INCOME_AMT_10000-24999  INCOME_AMT_100000-499999  \
0                0.0                     0.0                       0.0   
1                1.0                     0.0                       0.0   
2                0.0                     0.0                       0.0   
3                0.0                     1.0                       0.0   
4                0.0                     0.0                       1.0   

   INCOME_AMT_10M-50M  INCOME_AMT_1M-5M  INCOME_AMT_25000-99999  \
0                 0.0               0.0                     0.0   
1                 0.0               0.0                     0.0   
2                 0.0               0.0                     0.0   
3                 0.0               0.0                     0.0   
4                 0.0               0.0                     0.0   

   INCOME_AMT_50M+  INCOME_AMT_5M-10M  SPECIAL_CONSIDERATIONS_N  \
0              0.0                0.0                       1.0   
1              0.0                0.0                       1.0   
2              0.0                0.0                       1.0   
3              0.0                0.0                       1.0   
4              0.0                0.0                       1.0   

   SPECIAL_CONSIDERATIONS_Y  
0                       0.0  
1                       0.0  
2                       0.0  
3                       0.0  
4                       0.0  

[5 rows x 447 columns]

In [50]:
# Splitting preprocessed data into features and target arrays
y = application_df_copy['IS_SUCCESSFUL'].values
X = application_df_copy.drop(columns= 'IS_SUCCESSFUL').values

# Splitting the preprocessed data into a training and testing dataset
X_train, X_test, y_train, y_test = train_test_split(X, y, random_state=42, stratify=y)

# Creating a StandardScaler instance
scaler = StandardScaler()

# Fitting the StandardScaler
X_scaler = scaler.fit(X_train)

# Scaling the data
X_train_scaled = X_scaler.transform(X_train)
X_test_scaled = X_scaler.transform(X_test)

In [51]:
# Defining the model - deep neural net, i.e., the number of input features and hidden nodes for each layer.
number_input_features = len(X_train[0])

hidden_nodes_layer1 =  20
hidden_nodes_layer2 = 10

nn = tf.keras.models.Sequential()

# First hidden layer
nn.add(
    tf.keras.layers.Dense(units=hidden_nodes_layer1, input_dim=number_input_features, activation="relu")
)

# Second hidden layer
nn.add(tf.keras.layers.Dense(units=hidden_nodes_layer2, activation="relu"))

# Output layer
nn.add(tf.keras.layers.Dense(units=1, activation="sigmoid"))

# Checking the structure of the model
nn.summary()

Model: "sequential_3"
_________________________________________________________________
 Layer (type)                Output Shape              Param #   
 dense_11 (Dense)            (None, 20)                8940      
                                                                 
 dense_12 (Dense)            (None, 10)                210       
                                                                 
 dense_13 (Dense)            (None, 1)                 11        
                                                                 
Total params: 9161 (35.79 KB)
Trainable params: 9161 (35.79 KB)
Non-trainable params: 0 (0.00 Byte)
_________________________________________________________________


In [52]:
# Compiling the model
nn.compile(loss="binary_crossentropy", optimizer="adam", metrics=["accuracy"])

In [53]:
# Train the model
fit_model = nn.fit(X_train_scaled,y_train,epochs=20)

Epoch 1/20
804/804 [==============================] - 4s 3ms/step - loss: 0.5019 - accuracy: 0.7633
Epoch 2/20
804/804 [==============================] - 2s 3ms/step - loss: 0.4317 - accuracy: 0.7958
Epoch 3/20
804/804 [==============================] - 2s 3ms/step - loss: 0.4255 - accuracy: 0.7966
Epoch 4/20
804/804 [==============================] - 2s 3ms/step - loss: 0.4235 - accuracy: 0.7982
Epoch 5/20
804/804 [==============================] - 2s 3ms/step - loss: 0.4215 - accuracy: 0.7989
Epoch 6/20
804/804 [==============================] - 2s 3ms/step - loss: 0.4196 - accuracy: 0.8008
Epoch 7/20
804/804 [==============================] - 2s 3ms/step - loss: 0.4191 - accuracy: 0.8001
Epoch 8/20
804/804 [==============================] - 2s 3ms/step - loss: 0.4183 - accuracy: 0.8003
Epoch 9/20
804/804 [==============================] - 2s 3ms/step - loss: 0.4174 - accuracy: 0.8016
Epoch 10/20
804/804 [==============================] - 2s 3ms/step - loss: 0.4172 - accuracy: 0.8019

In [54]:
# Evaluating the model using the test data
model_loss, model_accuracy = nn.evaluate(X_test_scaled,y_test,verbose=2)
print(f"Loss: {model_loss}, Accuracy: {model_accuracy}")

268/268 - 1s - loss: 0.4428 - accuracy: 0.7941 - 699ms/epoch - 3ms/step
Loss: 0.44281700253486633, Accuracy: 0.7940524816513062


### A model accuracy of 79.4% is observed with model trained on as less as 20 epochs and 30 neurons between 3 layers with activation functions relu, relu and sigmoid. Loss has also reduced to 41%

In [55]:
# Saving weights of first layer with 20 neurons and 446 features at the end of final epoch
weights, biases = nn.layers[0].get_weights()
weights2 =pd.DataFrame(weights)
weights2.to_csv("Weights/checkweightmethod4.csv")

In [56]:
# Exporting model to HDF5 file
nn.save("Models/AlphabetSoupCharity_Optimisation.h5", save_format='h5')

C:\Users\jyots\anaconda3\envs\bootcamp3\lib\site-packages\keras\src\engine\training.py:3079: UserWarning: You are saving your model as an HDF5 file via `model.save()`. This file format is considered legacy. We recommend using instead the native Keras format, e.g. `model.save('my_model.keras')`.
  saving_api.save_model(
